tests/build_stitched_index.cpp

```diff
@@ -0,0 +1,855 @@
+                              number_of_label_points, dimension);
+    diskann::Index<T> index(diskann::Metric::L2, dimension,
+                            number_of_label_points, false, false);
+
+    auto index_build_timer = std::chrono::high_resolution_clock::now();
+    index.build(curr_label_input_data_path.c_str(), number_of_label_points,
+                label_index_build_parameters);
+    std::chrono::duration<double> current_indexing_time =
+        std::chrono::high_resolution_clock::now() - index_build_timer;
+
+    total_indexing_time += current_indexing_time.count();
+    indexing_percentage += (1 / (double) all_labels.size());
+    print_progress(indexing_percentage);
+
+    index.save(curr_label_index_path.c_str());
+  }
+  std::cout.clear();
+  diskann::cout.clear();
+
+  std::cout << "\nDone. Generated per-label indices in " << total_indexing_time
+            << " seconds\n"
+            << std::endl;
+}
+
+/*
+ * Manually loads a graph index in from a given file.
+ *
+ * Returns both the graph index and the size of the file in bytes.
+ */
+load_label_index_return_values load_label_index(path label_index_path,
+                                                _u32 label_number_of_points) {
+  std::ifstream label_index_stream;
+  label_index_stream.exceptions(std::ios::badbit | std::ios::failbit);
+  label_index_stream.open(label_index_path, std::ios::binary);
+
+  _u64         index_file_size, index_num_frozen_points;
+  _u32         index_max_observed_degree, index_entry_point;
+  const size_t INDEX_METADATA = 2 * sizeof(_u64) + 2 * sizeof(_u32);
+  label_index_stream.read((char *) &index_file_size, sizeof(_u64));
+  label_index_stream.read((char *) &index_max_observed_degree, sizeof(_u32));
+  label_index_stream.read((char *) &index_entry_point, sizeof(_u32));
+  label_index_stream.read((char *) &index_num_frozen_points, sizeof(_u64));
+  size_t bytes_read = INDEX_METADATA;
+
+  std::vector<std::vector<_u32>> label_index(label_number_of_points);
+  _u32                           nodes_read = 0;
+  while (bytes_read != index_file_size) {
+    _u32 current_node_num_neighbors;
+    label_index_stream.read((char *) &current_node_num_neighbors, sizeof(_u32));
+    nodes_read++;
+
+    std::vector<_u32> current_node_neighbors(current_node_num_neighbors);
+    label_index_stream.read((char *) current_node_neighbors.data(),
+                            current_node_num_neighbors * sizeof(_u32));
+    label_index[nodes_read - 1].swap(current_node_neighbors);
+    bytes_read += sizeof(_u32) * (current_node_num_neighbors + 1);
+  }
+
+  return std::make_tuple(label_index, index_file_size);
+}
+
+/*
+ * Custom index save to write the in-memory index to disk.
+ * Also writes required files for diskANN API -
+ *  1. labels_to_medoids
+ *  2. universal_label
+ *  3. data (redundant for static indices)
+ *  4. labels (redundant for static indices)
+ */
+void save_full_index(path final_index_path_prefix, path input_data_path,
```

Comment:
Would we want this here or in the Index class as a save_filtered_index method?

Reply:
The index class already essentially has a "save_filtered_index" method. However, since the graph stitching occurs outside of the index API (via this file's `stitch_label_indices` function), the final result cannot be saved using that function. So, we simply redefine it here.